### PR TITLE
Detect effective side as SERVER when an EpollEventLoopGroup is being used

### LINKED ISSUE #3633: FMLCommonHandler.instance().effectiveSide returns Side.CLIENT on Linux server
Problematic line: https://github.com/MinecraftForge/MinecraftForge/blob/1.11.x/src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java#L205

Everything works as intended on my Windows machine but my Linux server has differently named Netty threads:
Threads Screenshot

Fix should be as simple as adding another clause to the if statement in FMLCommonHandler.

### PR DESCRIPTION
**Pull request by kashike**
Fixes #3633

### CODE REVIEW

File: src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java [hunk 1]
```diff
@@ -202,7 +202,7 @@ public Side getSide()
     public Side getEffectiveSide()
     {
         Thread thr = Thread.currentThread();
-        if (thr.getName().equals("Server thread") || thr.getName().startsWith("Netty Server IO"))
+        if (thr.getName().equals("Server thread") || thr.getName().startsWith("Netty Epoll Server IO") || thr.getName().startsWith("Netty Server IO"))
```
Comment (bs2609) [inline src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java]: Just use `thr.getName().contains("Server")`

Comment (diesieben07) [reply src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java]: No, that might lead to things being detected as the Minecraft server that are some other kind of server

Comment (bs2609) [reply src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java]: Such as?
Checking only for exact names is the reason for this pull request existing.
If accuracy is important, this method is the wrong way to approach the problem anyway.

Comment (kashike) [reply src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java]: Mods can include dependencies which start threads with names containing "Server", or a mod can start a thread with a name containing "Server" itself. `getEffectiveSide` is *not* meant to cover all cases.

Comment (bs2609) [reply src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java]: My point was, as it's not going to work properly in all cases regardless, why not keep it simple? Is misidentifying as client particularly preferable to misidentifying as server?

**Comment by mezz** [comment]
What are the valid uses for `getEffectiveSide()`? It seems like something that should be really discouraged.

**Comment by raoulvdberge** [comment]
It should be discouraged, but it's broken, so it should be fixed.

**Comment by mezz** [comment]
That's not the question... of course it should be fixed.

**Comment by diesieben07** [comment]
Two things I can think of:
- Logging
- Safety checks, i.e. checking that a method that is only supposed to be called on the client thread is really only called there.

**Comment by mezz** [comment]
Looking more into it, it seems like it's mainly being used to detect when you're on single player? FML uses it to handle a couple sided quirks deep in vanilla that mods do not have to deal with.

Can you explain how it is useful for logging?
For safety checks... maybe in dev but it seems like a bad idea for performance.

The only other use I can think up is for bypassing network sync when in single player, which is not a very safe optimization to try to pull off, with very little benefit.

@raoulvdberge I assume you use it, what do you need it for?

**Comment by diesieben07** [comment]
Using `getEffectiveSide` to detect single player seems counter-intuitive to me. Single player is something that can only happen on the _physical_ client, so not sure why you'd use `getEffectiveSide`.

As for logging, you might want to have a generic log method that logs the current logical side. But thinking about it, FML already appends the thread name, so maybe this is not as useful.

Safety checks, yes, definitely something only to be used in dev, guarded by a static check, so that the JVM can optimize it away outside of dev.

As for the network sync, I did that in one case (because the data was several MB in some cases, so it was worth it) and I did not need to use `getEffectiveSide`. Detecting if the player is the single-player is done via `getServerOwner`.

**Comment by mezz** [comment]
I spent some time looking over it with diesieben07 and we found that there is no way to fully replace `getEffectiveSide`. FML has some uses for it to handle special cases in the vanilla client/server split that can't be handled any other way.
It's still recommended that mods try hard to use less hacky alternatives wherever they can.

**Comment by liach** [comment]
@mezz Can we replace it with things like `isOnServerThread` instead of returning a confusing enum?

**Comment by LatvianModder** [comment]
What happened to the idea of assigning ThreadGroups to make this safer and faster? Something like - on init, tell which threads are client only, and after that all new threads created by those will also be marked as client, the rest will be server

**Comment by bs2609** [comment]
I suggested it in #3279.

**Comment by diesieben07** [comment]
That route sounds much cleaner to me, it's the exact use-case for thread groups.

**Pull request merged**